### DISCUSSION TITLE
ICancellationTokenSource

### DISCUSSION THREAD

**Discussion by timcassell** [dotnet/runtime#54460] [Ideas] [answered]
In the same vein as `IValueTaskSource`, I would like `CancellationToken` to be updated to allow an `ICancellationTokenSource` passed to a new constructor.

Reason: In my [ProtoPromise](https://github.com/timcassell/ProtoPromise) library, I have a custom `CancelationSource` which is a struct backed by a class that gets pooled for reuse when it is disposed. As it is now, `System.Threading.CancellationTokenSource` is a class that cannot be reused (much like `TaskCompletionSource`), and thus subject to garbage collection.

Async iterators added `System.Threading.CancellationToken` as part of the contract, so I would like to be able to freely convert my custom `CancelationToken` to `System.Threading.CancellationToken` without having to allocate a new `System.Threading.CancellationTokenSource`.

**Comment by stephentoub** [comment, marked as answer]
Thanks for the suggestion, but we've discussed this before and don't want to do it.  CancellationToken has a strong guarantee that once it transitions to having cancellation requested,  it'll never transition back,  and such use would break that guarantee.  You can,  of course,  reuse a CTS as many times as you'd like if it hasn't yet had cancellation requested (and if it has, the associated costs almost always are orders of magnitude larger than what you'd potentially save by pooling),  subject to whatever guarantees your own code provides,  and we added TryReset in .NET 6 to make that easier,  in particular in the face of timeouts.

**Reply by timcassell** [reply]
@stephentoub I see, that makes a lot of sense. Certainly the new `TryReset` should make it much better than allocating new every time, especially since cancelations are the exception, not the rule.

Even better would be the ability to use custom cancelation tokens with async iterators, but I can't imagine a clean way of doing that. :stuck_out_tongue: 

In case you were wondering, my own `CancelationToken` has a `(Try)Retain`/`Release` for the consumer to protect its integrity for the duration it's needed, though in practice it's generally not needed if the producer only disposes when the requested operation completes/fails.

Thanks for the response!
